### PR TITLE
IOBalancer Tests Added and minor change in Connection Registration in IOBalancer

### PR DESCRIPTION
**Pull request by jerrinot**
1. IOBalancer Tests Added
2. Refactoring: - IOBalancer is now a top-level class
3. Different integration point with TcpIpConnectionManager

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java [hunk 1]
```diff
@@ -77,7 +75,7 @@ public IOBalancer(InSelectorImpl[] inSelectors, OutSelectorImpl[] outSelectors,
         this.inLoadTracker = new LoadTracker(inSelectors, loggingService);
         this.outLoadTracker = new LoadTracker(outSelectors, loggingService);
 
-        this.isActive = shouldStart(inSelectors, outSelectors);
+        this.shouldStart = shouldStart(inSelectors, outSelectors);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: I like the name active a lot more.

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java [hunk 2]
```diff
@@ -178,26 +179,4 @@ private void tryMigrate(LoadImbalance loadImbalance) {
         }
         handler.migrate(destinationSelector);
     }
-
-    private final class IOBalancerThread extends Thread {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: Why did you move the thread to the outside? It does almost nothing and it relies on the internals of the balancer.

File: hazelcast/src/test/java/com/hazelcast/nio/tcp/handlermigration/LoadTrackerTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        loadTracker.updateImbalance();
+        assertEquals(100, loadImbalance.minimumEvents);
+        assertEquals(300, loadImbalance.maximumEvents);
+        assertEquals(selector1, loadImbalance.destinationSelector);
+        assertEquals(selector2, loadImbalance.sourceSelector);
+    }
+
+    @Test
+    public void testAddHandler() throws Exception {
+
```
Comment (pveentjer) [inline hazelcast/src/test/java/com/hazelcast/nio/tcp/handlermigration/LoadTrackerTest.java]: Ahh.. the famous empty line test :)

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java (hunk 1 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: me too, but it's misleading. 
it's set to `true` in a constructor even when it's not active yet. it's active only after calling the `start()` method.

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java (hunk 2 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: the original idea was to improve testability. I didn't help great deal, but I like it outside anyway.

File: hazelcast/src/test/java/com/hazelcast/nio/tcp/handlermigration/LoadTrackerTest.java (hunk 3 above)
Comment (jerrinot) [inline hazelcast/src/test/java/com/hazelcast/nio/tcp/handlermigration/LoadTrackerTest.java]: ops, good catch, thanks!

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java (hunk 1 above)
Comment (noctarius) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: Isn't it enabled and running, started or active?

Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: `enabled` is a good one!

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java [hunk 4]
```diff
@@ -77,7 +75,7 @@ public IOBalancer(InSelectorImpl[] inSelectors, OutSelectorImpl[] outSelectors,
         this.inLoadTracker = new LoadTracker(inSelectors, loggingService);
         this.outLoadTracker = new LoadTracker(outSelectors, loggingService);
 
-        this.isActive = shouldStart(inSelectors, outSelectors);
+        this.enabled = shouldStart(inSelectors, outSelectors);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: I would call the method: isEnabled.

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java (hunk 2 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancer.java]: I'm fine with leaving since it a flavor thing. But the thread is small and pokes in the internal of that class. The thread is tied to the implementation, so I make it internal. Unless it becomes big, then I rip it apart.

File: hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancerThread.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        this.ioBalancer = ioBalancer;
+        this.log = log;
+        this.migrationIntervalSeconds = migrationIntervalSeconds;
+    }
+
+    @Override
+    public void run() {
+        try {
+            log.finest("Starting IOBalancer thread");
+            while (!Thread.interrupted()) {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/IOBalancerThread.java]: I never rely on interrupted as flag for a thread to stop. In your case it is ok since you know the InterruptedException isn't swallowed. But it can turn unpredictable quickly.

**Comment by jerrinot** [comment]
verify
